numpyro/distributions/continuous.py

```diff
@@ -290,6 +291,62 @@ def infer_shapes(concentration):
         return batch_shape, event_shape
 
 
+class EulerMaruyama(Distribution):
+    arg_constraints = {"t": constraints.ordered_vector}
+
+    support = constraints.real
+
+    def __init__(self, t, sde_fn, sde_pars, init_dist, validate_args=None):
+        # we use t rather than (dt, num_steps) to make the api simpler and more general,
+        # `init_dist` will define prior for the init value,
+        # event dimensions will be decided by `init_dist`
+        self.t = t
+        self.dt = jnp.diff(t)
+        self.sde_fn = sde_fn
+        self.sde_pars = sde_pars
+        self.init_dist = init_dist
+
+        batch_shape, event_shape = init_dist.batch_shape, jnp.shape(t)
+
+        super(EulerMaruyama, self).__init__(
+            batch_shape, event_shape, validate_args=validate_args
+        )
+
+    def sample(self, key, sample_shape=()):
+        assert is_prng_key(key)
+        noises = random.normal(key, shape=sample_shape + self.event_shape)
+
+        def step(y_curr, xs):
+            t_curr, dt_curr, noise_curr = xs[0], xs[1], xs[2:].T
+
+            f, g = self.sde_fn(y_curr, t_curr, self.sde_pars)
+            mu = y_curr + dt_curr * f
+            sigma = jnp.sqrt(dt_curr) * g
+            y_next = mu + sigma * noise_curr
+            return y_next, y_next
+
+        xs = jnp.hstack(
+            (self.t[..., :-1, jnp.newaxis], self.dt[:, jnp.newaxis], noises[..., :-1].T)
+        )
+
+        _, sde_out = scan(
+            step, self.init_dist.sample(key, sample_shape=sample_shape), xs
```

Comment:
you can assume `sample_shape = ()` for now. For arbitrary sample_shape, you need to do something like
```
inits = self.init_dist.sample(key, sample_shape=sample_shape)
vmap_axes = list(range(len(sample_shape)))
_, sde_out = vmap(lambda init, noise: scan(step, init, (noise, self.t[:-1], self.dt)), vmap_axes, vmap_axes)(inits, noises)
cat_dim = -self.event_dim - 1
sde_out = jnp.concatenate([jnp.expand_dims(inits, cat_dim), sde_outs], axis=cat_dim)
```

Reply:
implemented assuming arbitrary sample_shape